test/kafka-upsert-benchmark/mzcompose.py

```diff
@@ -0,0 +1,176 @@
+    query = "SELECT * FROM num_records_ingested"
+
+    try:
+        conn = pg8000.connect(
+            host="localhost", port=host_port, user="materialize", timeout=1
+        )
+        conn.autocommit = True
+
+        cursor = conn.cursor()
+        cursor.execute(query)
+        n = 0
+        result = cursor.fetchone()[0]
+        if result is not None:
+            n = int(result)
+        return n
+    except Exception as e:
+        raise UIError(f"error querying materialize: {e}")
+
+
+def send_records(
+    c: Composition,
+    num_records: int,
+    num_keys: int,
+    value_bytes: int,
+    timeout_secs: int,
+) -> None:
+    c.run(
+        "kgen",
+        f"--num-records={num_records}",
+        "--keys=random",
+        "--key-min=0",
+        f"--key-max={num_keys}",
+        "--values=bytes",
+        f"--max-message-size={value_bytes+1}",
+        f"--min-message-size={value_bytes}",
+        "--topic=testdrive-load-test-1",
+        "--quiet",
+    )
+
+
+# This workflow runs an open loop benchmark where it tries to send a steady rate
+# of records to Kafka, and tracks how many records Materialize has ingested up to.
+# Crucially, the rate of insertion into Kafka is as independent as possible from the
+# rate of those records are ingested into Materialize, to try to avoid the
+# "coordinated ommission" problem [1].
+#
+# Repeated runs of this benchmark can answer the question "what's the peak messages/sec
+# Materialize can ingest for this workload" where the "workload" could be "persisted
+# upsert Kafka sources with 10 MM unique keys and 2kb values", and comparing the peak QPS
+# observed across different such workloads (along with the corresponding dataflow timing
+# info) can give insight into the relative overheads of different workloads and what
+# might be causing them.
+#
+# [1]: https://www.scylladb.com/2021/04/22/on-coordinated-omission/
+def workflow_kafka_upsert_benchmark(
+    c: Composition, parser: WorkflowArgumentParser
+) -> None:
+    parser.add_argument("--num-seconds", type=int, default=100)
+    parser.add_argument("--records-per-second", type=int, default=10000)
+    parser.add_argument("--num-keys", type=int, default=1000000000)
+    parser.add_argument("--value-bytes", type=int, default=500)
+    parser.add_argument("--timeout-secs", type=int, default=120)
+    parser.add_argument("--enable-persistence", action="store_true")
+    args = parser.parse_args()
+
+    c.start_and_wait_for_tcp(services=prerequisites)
+    mz_service = "materialized"
+    if args.enable_persistence:
+        mz_service = "materialized_persist"
+
+    c.up(mz_service)
+    c.wait_for_materialized(mz_service)
+
+    c.run(
+        "testdrive-svc",
+        "setup.td",
+    )
+
+    start = time.monotonic()
+    records_sent = 0
+    # Maximum observed delta between records sent by the benchmark and ingested by
+    # Materialize.
+    max_lag = 0
+    last_reported_time = 0.0
+    say(
```

Comment:
@benesch I'm going to need your help debugging this. I needed this println because without it the benchmark just silently quits while trying to find the host port. Most printlns I add make the issue go away though!